### PR TITLE
Adding new pipeline element `fill_to`

### PR DESCRIPTION
**Pull request by WireBaron**
This new element will take a method type and an interval.  It will then
scan through the passed timevector for any gaps larger than the given
interval, and will fill them according to the method type.

### CODE REVIEW

File: extension/src/time_series/pipeline/fill_to.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        SeriesType::NormalSeries{step_interval, ..} => if step_interval <= interval { return series.clone(); },
+        _ => ()
+    }
+
+    let mut result: Option<Vec<TSPoint>> = None;
+    let mut it = series.iter().peekable();
+    let mut current = it.next();
+
+    while current.is_some() {
+        let lhs = current.unwrap();
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:94]:
```suggestion
    while let (lhs) = current {
```

File: extension/src/time_series/pipeline/fill_to.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+            if result.is_none() {
+                result = Some(series.iter().collect());
+            }
+            let result = result.as_mut().unwrap();
+            let mut target = lhs.ts + interval;
+            while target < rhs.ts {
+                result.push(method.fill_point(&lhs, &rhs, target));
+                target += interval;
+            }
+        }
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:108]:
```suggestion
        match it.peek() {
            Some(rhs) if rhs.ts - lhs.ts > interval => {
                let result = match &mut result {
                    Some(result) => result,
                    None => {
                        *result = Some(series.iter().collect());
                        result.as_mut().unwrap()
                    },
                }
                let mut target = lhs.ts + interval;
                while target < rhs.ts {
                    result.push(method.fill_point(&lhs, &rhs, target));
                    target += interval;
                }
            }
            _ => {},
        }
```

File: extension/src/time_series/pipeline/fill_to.rs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        let lhs = current.unwrap();
+        let rhs = it.peek();
+
+        if rhs.is_some() && rhs.unwrap().ts - lhs.ts > interval {
+            let rhs = rhs.unwrap();
+            if result.is_none() {
+                result = Some(series.iter().collect());
+            }
+            let result = result.as_mut().unwrap();
+            let mut target = lhs.ts + interval;
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:103]: we should add a
```
// FIXME use postgres interval ops if the interval contains anything larger than an hour
```

File: extension/src/time_series/pipeline/fill_to.rs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+        current = it.next();
+    }
+
+    if result.is_some() {
+        let mut result = result.unwrap();
+        result.sort_unstable_by_key(|p| p.ts);
+        build!{
+            Timevector {
+                series: SeriesType::ExplicitSeries{
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:118]:
```suggestion
                series: SeriesType::SortedSeries{
```

File: extension/src/time_series/pipeline/fill_to.rs [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+                target += interval;
+            }
+        }
+
+        current = it.next();
+    }
+
+    if result.is_some() {
+        let mut result = result.unwrap();
+        result.sort_unstable_by_key(|p| p.ts);
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:115]: 1. do we need this? is this guaranteed to be sorted by the fact that we have sorted input?
2. Unless we deduplicate, I think we should use a stable sort, that we if there's meaning to the current order we'll preserve it
```suggestion
        result.sort_by_key(|p| p.ts);
```

Comment (JLockerman) [reply extension/src/time_series/pipeline/fill_to.rs:115]: > is this guaranteed to be sorted...
Oh, I see: we push the new elements onto the end of the vector

File: extension/src/time_series/pipeline/fill_to.rs [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            let result = result.as_mut().unwrap();
+            let mut target = lhs.ts + interval;
+            while target < rhs.ts {
+                result.push(method.fill_point(&lhs, &rhs, target));
+                target += interval;
+            }
+        }
+
+        current = it.next();
+    }
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:109]: Just for fun, if we were only working on slices we could do
```suggestion
    let mut result = vec![];
    let mut windows = series.points.windows(2);

    for window in windows {
        if window[1].ts - window[0].ts > interval {
            let result = result.as_mut().unwrap();
            let mut target = window[0].ts + interval;
            while target < window[1].ts {
                result.push(method.fill_point(&window[0], &window[1], target));
                target += interval;
            }
        }
    }
```

File: extension/src/time_series/pipeline/fill_to.rs (hunk 3 above)
Comment (WireBaron) [reply extension/src/time_series/pipeline/fill_to.rs]: added above

File: extension/src/time_series/pipeline/fill_to.rs (hunk 1 above)
Comment (WireBaron) [reply extension/src/time_series/pipeline/fill_to.rs]: nice, knew there had to be a better way to do this.

File: extension/src/time_series/pipeline/fill_to.rs (hunk 2 above)
Comment (WireBaron) [reply extension/src/time_series/pipeline/fill_to.rs]: much nicer!

File: extension/src/time_series/pipeline/fill_to.rs (hunk 4 above)
Comment (WireBaron) [reply extension/src/time_series/pipeline/fill_to.rs]: Ooops, that's an important fix

File: extension/src/time_series/pipeline/fill_to.rs (hunk 5 above)
Comment (WireBaron) [reply extension/src/time_series/pipeline/fill_to.rs]: fixed

File: extension/src/time_series/pipeline/fill_to.rs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        match it.peek() {
+            Some(rhs) if rhs.ts - lhs.ts > interval => {
+                let mut target = lhs.ts + interval;
+                while target < rhs.ts {
+                    result.push(method.fill_point(&lhs, &rhs, target));
+                    target += interval;
+                }
+            },
+            _ => {},
+        }
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:101]:
```suggestion
    while let (Some(lhs), Some(rhs) = (current, it.peek()) {
        if rhs.ts - lhs.ts > interval => {
            let mut target = lhs.ts + interval;
            while target < rhs.ts {
                result.push(method.fill_point(&lhs, &rhs, target));
                target += interval;
            }
        }
```

File: extension/src/time_series/pipeline.rs [hunk 8]
```diff
@@ -176,6 +186,8 @@ pub fn execute_pipeline_element<'s, 'e>(
             return filter::apply_lambda_to(timevector, lambda),
         Element::Arithmetic{ function, rhs } =>
             return arithmetic::apply(timevector, *function, *rhs),
+        Element::FillTo{..} =>
+            return fill_to(&timevector, &element),
```
Comment (JLockerman) [inline extension/src/time_series/pipeline.rs:190]:
```suggestion
            return fill_to(timevector, &element),
```

File: extension/src/time_series/pipeline/fill_to.rs [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+        Element::FillTo {
+            interval,
+            fill_method
+        }.flatten()
+    }
+}
+
+pub fn fill_to<'s>(
+    series: &toolkit_experimental::Timevector<'s>,
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:73]:
```suggestion
    series: toolkit_experimental::Timevector<'s>,
```

File: extension/src/time_series/pipeline/fill_to.rs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+            Timevector {
+                series: SeriesType::SortedSeries{
+                    num_points: result.len() as _,
+                    points: result.into(),
+                }
+            }
+        }
+    } else {
+        series.clone()
+    }
```
Comment (JLockerman) [inline extension/src/time_series/pipeline/fill_to.rs:119]:
```suggestion
    
    if result.is_empty() {
        return series
    }
            let mut result: Vec<TSPoint> = series.iter().chain(result.into_iter()).collect();
        result.sort_by_key(|p| p.ts);
        build!{
            Timevector {
                series: SeriesType::SortedSeries{
                    num_points: result.len() as _,
                    points: result.into(),
                }
            }
        }

```

**Pull request merged**